src/agent/onefuzz/src/monitor.rs

```diff
@@ -49,48 +55,60 @@ impl DirectoryMonitor {
         Ok(())
     }
 
-    pub fn poll_file(&mut self) -> Poll<Option<PathBuf>> {
-        match self.rx.try_recv() {
-            Ok(DebouncedEvent::Create(path)) => Poll::Ready(Some(path)),
-            Ok(DebouncedEvent::Remove(path)) => {
-                if path == self.dir {
-                    // The directory we were watching was removed; we're done.
-                    self.stop().ok();
-                    Poll::Ready(None)
-                } else {
-                    // Some file _inside_ the watched directory was removed.
-                    Poll::Pending
-                }
-            }
-            Ok(_evt) => {
-                // Filesystem event we can ignore.
-                Poll::Pending
-            }
-            Err(TryRecvError::Empty) => {
-                // Nothing to read, but sender still connected.
-                Poll::Pending
+    pub async fn next_file(&mut self) -> Option<PathBuf> {
+        loop {
+            let event = self.notify_events.recv().await;
+
+            if event.is_none() {
+                // Make sure we stop our `Watcher` if we return early.
+                let _ = self.stop();
             }
-            Err(TryRecvError::Disconnected) => {
-                // We'll never receive any more events; whatever happened, we're done.
-                self.stop().ok();
-                Poll::Ready(None)
+
+            match event? {
+                DebouncedEvent::Create(path) => {
+                    return Some(path);
+                }
+                DebouncedEvent::Remove(path) => {
+                    if path == self.dir {
+                        // The directory we were watching was removed; we're done.
+                        let _ = self.stop();
+                        return None;
+                    } else {
+                        // Some file _inside_ the watched directory was removed. Ignore.
+                    }
+                }
+                _event => {
+                    // Other filesystem event. Ignore.
+                }
             }
         }
     }
 }
 
-impl Stream for DirectoryMonitor {
-    type Item = PathBuf;
+fn into_async<T: Send + 'static>(
```

Comment:
Document, though internal. Note that the `JoinHandle` is inessential, and we could just drop it.

Thought: in doesn't matter in our usage here, but maybe we should use `tokio::task::spawn_blocking()` and return its `JoinHandle`, instead. By using `thread::spawn()`, we return a `JoinHandle` whose `join()` method is sync and blocks the async runtime, which means we'll almost certainly never want to use it.

Reply:
It would make the code more consistent, but the impact is minimal.

This has no mechanism for communicating errors beyond having the channel close.  From the `DirectoryMonitor` callers perspective, this looks the same as the directory being deleted.

Reply:
Oh, only just now saw this comment. I changed it anyway because it literally only required renaming.